Template type checking support for @let

crisbeto opened the pull request:
These changes include the code necessary to type check `@let` declarations. Also some prerequisites for the language service integration.

### refactor(compiler-cli): add compiler flag for testing let declarations 
Adds a private `_enableLetSyntax` flag that allows for let declarations to be enabled in tests.

### refactor(compiler): integrate let declarations into the template binder
Integrates the let declarations into the template binder which will be used to power the scoping behavior of the new syntax.

### refactor(compiler-cli): integrate let declarations into the template type checker
Integrates let declarations into the template type checker by producing corresponding constants in the TCB.
This also includes a couple of custom diagnostics to flag usages of let before they're declared and illegal writes to let declarations. We can't rely on TS for these checks, because it includes the variable name in the diagnostic.

### refactor(compiler-cli): integrate let declaration into the indexer
Adds support for let declarations in the template indexer.

### refactor(compiler-cli): support resolving the symbol of let declaration
Updates the symbol builder to handle resolving the symbol of a let declaration.

### refactor(compiler-cli): support completions for let declarations
Adds support for let declarations inside the `CompletionEngine`.

### refactor(compiler-cli): account for let declarations in two-way binding check
Updates the check that prevent writes to template variables in two-way bindings to account for let declarations.
Also fixes some old tests that weren't properly setting up two-way bindings.

### refactor(compiler-cli): add diagnostic for duplicate let declarations
Adds a template diagnostic that will flag cases where multiple `@let` declarations use the same name.

crisbeto committed:
refactor(compiler-cli): add compiler flag for testing let declarations

Adds a private `_enableLetSyntax` flag that allows for let declarations to be enabled in tests.

crisbeto committed:
refactor(compiler): integrate let declarations into the template binder

Integrates the let declarations into the template binder which will be used to power the scoping behavior of the new syntax.

crisbeto committed:
refactor(compiler-cli): integrate let declarations into the template type checker

Integrates let declarations into the template type checker by producing corresponding constants in the TCB.

This also includes a couple of custom diagnostics to flag usages of let before they're declared and illegal writes to let declarations. We can't rely on TS for these checks, because it includes the variable name in the diagnostic.

crisbeto committed:
refactor(compiler-cli): integrate let declaration into the indexer

Adds support for let declarations in the template indexer.

crisbeto committed:
refactor(compiler-cli): support resolving the symbol of let declaration

Updates the symbol builder to handle resolving the symbol of a let declaration.

crisbeto committed:
refactor(compiler-cli): support completions for let declarations

Adds support for let declarations inside the `CompletionEngine`.

crisbeto committed:
refactor(compiler-cli): account for let declarations in two-way binding check

Updates the check that prevent writes to template variables in two-way bindings to account for let declarations.

Also fixes some old tests that weren't properly setting up two-way bindings.

crisbeto commented on packages/compiler/src/render3/view/t2_api.ts:
```diff
@@ -50,6 +51,9 @@ export type ReferenceTarget<DirectiveT> =
   | Template;
 
+/** Entity that is local to the template and defined within the template. */
+export type TemplateEntity = Reference | Variable | LetDeclaration;
```
Context for this type is that we were repeating `Reference | Variable` in a bunch of places which would now have to include `LetDeclaration`. I moved it out to a type to make it easier to manage.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -360,6 +370,34 @@ class TcbTemplateContextOp extends TcbOp {
 }
 
+/**
+ * A `TcbOp` which generates a variable for a `TmplAstTemplate`'s context.
```
This comment is in need of some adjustments

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -2633,18 +2695,36 @@ class TcbExpressionTranslator {
-      return this.resolveTarget(ast);
+      const target = this.tcb.boundTarget.getExpressionTarget(ast);
+      if (target instanceof TmplAstLetDeclaration) {
+        this.validateLetDeclarationAccess(target, ast);
+        // Let declarations can't be accessed through `this`.
+        if (ast.receiver instanceof ThisReceiver) {
+          return null;
+        }
```
The validation seems a bit odd to me if we eventually bail because of an explicit `this`.

This is ultimately a bug in the binder (it shouldn't resolve explicit `this` accesses to local entities: https://github.com/angular/angular/pull/55183#issuecomment-2054157545). Maybe we can fix it there for `@let` declarations specifically, as the wider fix was breaking.

dylhunn replied:
@JoostK I have almost fixed all the breakages, and I'm about to land the other PR.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -2749,21 +2830,27 @@ class TcbExpressionTranslator {
+    const targetEnd = target.sourceSpan.end.offset;
+    const astStart = ast.sourceSpan.start;
+
+    // We only flag local references that occur before the declaration, because embedded views
+    // are updated before the child views. In practice this means that something like
+    // `<ng-template [ngIf]="true">{{value}}</ng-template> @let value = 1;` is valid.
```
Do we _want_ to allow this? I feel like I'd rather not allow this, at least initially, unless there's some use-cases that benefit from this ability?

crisbeto replied:
I initially had it set up to flag cases like this as well, but I changed it after talking with @alxhub about it. My thinking is:
1. It matches how local references work so it would be more familiar for users.
2. There's no technical reason why we can't support it since the value is guaranteed to be defined before child views attempt to reference it. There's one edge case where they can read an un-initialized variable, but I have safeguards against it in a future commit.
3. At least to me, it makes sense to "hoist" it, if we treat embedded views as JS functions.

Also worth noting that `@let` will be in developer preview initially and I'll keep an eye on feedback around this

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -2633,18 +2695,36 @@ class TcbExpressionTranslator {
+      // Ignore diagnostics from TS produced for writes to `@let` and re-report them using
+      // our own infrastructure. We can't rely on the TS reporting, because it includes
+      // the name of the auto-generated TCB variable name.
+      if (target instanceof TmplAstLetDeclaration) {
```
This should also take an explicit `this` into account, unless the binder is adjusted.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -3089,6 +3176,11 @@ class TcbEventHandlerTranslator extends TcbExpressionTranslator {
   }
+
+  protected override validateLetDeclarationAccess(): void {
+    // Event listeners are allowed to read `@let` declarations before
+    // they're declared since the callback won't be executed immediately.
```
Similarly here, is there real use-cases where this is useful?

crisbeto replied:
I think that this is basically the same discussion as the one for embedded views above so I'll resolve this.

JoostK commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -6609,5 +6609,447 @@ suppress
+          @Component({
+            template: \`
+              @let value = {} + 1;
+              {{value}}
```
Consider removing this usage to ensure the let expression is also type-checked if not used.

JoostK commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -6609,5 +6609,447 @@ suppress
+        expect(diags.length).toBe(0);
+      });
+
+      it('should not allow a let declaration value to be changed', () => {
```
Consider having a similar test with an explicit `this` access.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/api/symbols.ts:
```diff
@@ -242,6 +245,32 @@ export interface VariableSymbol {
+  /** The `ts.Type` of the entity. */
+  tsType: ts.Type;
+
+  /** The `ts.Symbol` for the declaration. */
```
Could you include what `null` means here.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/api/symbols.ts:
```diff
@@ -242,6 +245,32 @@ export interface VariableSymbol {
+  localVarLocation: TcbLocation;
+
+  /**
+   * The location in the shim file of the `@let` declaration.
```
nit:

```suggestion
   * The location in the shim file of the `@let` declaration's initializer expression.
```

JoostK reviewed (commented):
I haven't seen any diagnostics for duplicate `@let` declaration identifiers in the same scope.

crisbeto committed:
refactor(compiler-cli): add diagnostic for duplicate let declarations

Adds a template diagnostic that will flag cases where multiple `@let` declarations use the same name.

crisbeto commented:
@JoostK I've addressed the feedback and added another commit for duplicate declarations.

josephperrott reviewed (approved):
LGTM

atscott reviewed (approved):
reviewed-for: language-service

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

alxhub reviewed (approved):
Reviewed-for: public-api

thePunderWoman reviewed (approved):
reviewed-for: public-api

crisbeto commented:
Caretaker note: there's one pre-existing broken target that can be ignored.
